### ISSUE TITLE
Loss is constant

### ISSUE THREAD

**Issue by tarunn2799** [mlfoundations/open_clip#17]
I'm using CLIP to train on my custom dataset with the following params: 

Dataset size : 50k image-text pairs
Batch size : 128
Image Size : 224
Gpus : 1 
Epochs : 500

It's been running for a while now, I'm on my 15th epoch, and the loss hasn't changed at all. It isn't a constant number, but its constantly at 4.8xxx. Should I be concerned? I'm not sure why this is happening. 

![image](https://user-images.githubusercontent.com/28048963/133154185-01bdd63f-b3bc-460b-a583-21f5f9616a02.png)

**Comment by carlini** [comment]
This definitely looks like it's not training properly. One thing you could try is to reduce your dataset size to something tiny (128 images) and check if you can overfit those images. If that doesn't work then it would indicate some serious issue. Another thing you could try is to make sure that you can train on conceptual captions and the loss decreases there.

**Comment by tarunn2799** [comment]
Yes it does reduce normally when I'm just training with 1000 images. I can see a decrease at every epoch. How do I overcome this issue?

**Comment by carlini** [comment]
Without any additional information it's kind of hard to answer. If the loss decreases on small data, then the code is likely working correctly. It'd start with all the standard tests for working with a new dataset. Can you fit a really simple model to solve the task (at least partially, maybe even DNN on pixels + BoW on text)? Or is the data just too ugly to prevent learning on it entirely? Does tuning hyperparameters help at all? What happens if you train on increasingly larger datasets? Does validation loss decrease if you train on small training datasets? What do the gradients look like?

**Comment by tarunn2799** [comment]
One thing I did forget to mention is that I reduced the warmup to 1, since I thought the default value present (10000) was hardcoded for the conceptual captions dataset. I'm a little new to this. What would you recommend I change the warmup value to?

And yeah that's what I thought too, but it looks like its saturating around 4.85 again.  To answer your questions:
The data is actually pretty bad to use with simple models. 
I shall try out running the code to answer the remaining questions and update here. Thank you for your time.

**Comment by carlini** [comment]
I'd put the warmup back to 10000 (or a few thousand). This is an important piece of the code to make sure the loss doesn't explode which might explain some of what you're seeing.

**Comment by caiminghong** [comment]
Hi, carlini, I also encountered the sample problem as @tarunn2799 when train Vit -B/ 32. The loss function is implemented as follows:
def get_loss(logits_per_image, logits_per_text):
    labels = torch.eye(logits_per_image.size()[0]).cuda()
    loss_text = cross_entropy(logits_per_text, labels, reduction='none')
    loss_image = cross_entropy(logits_per_image, labels, reduction='none')

    loss = (loss_text + loss_image) / 2.0

    return loss.mean()

The training loss has been fluctunating around 7.277 when train the model with Flicker8k, and it seems that there is no training. The loss of the validation dataset is always 4.85. How should I analyze this problem? Thanks very much.

**Comment by mitchellnw** [comment]
Hi @caiminghong, two quick questions:
1. Does the model train for you on a dataset such as CC3M?
2. Is there any reason for using a different `get_loss` than [the one in the repo](https://github.com/mlfoundations/open_clip/blob/main/src/training/train.py#L23)? For instance, I think `cross_entropy` is not expecting a one-hot vector, e.g., we use `torch.arange(...)` instead of `torch.eye(...)`. Could be wrong here.

**Comment by carlini** [comment]
Trying to understand -- is this a new loss function you've modified? (If so, what happens if you make it the normal loss function?) And if not, can you also check if normal training works before trying a new dataset?

**Comment by carlini** [comment]
Sorry let's back up one step. You say "I just built the training process": are you using the code we've written in this repository? Or did you write your own version of CLIP training?

**Comment by rwightman** [comment]
closing as this is likely stale and current loss impl here as gone through a lot of testing lately without issues.
